.github/actions/setup/action.yml

```diff
@@ -8,31 +8,28 @@ inputs:
 runs:
   using: composite
   steps:
-    - name: Checkout all commits
-      uses: actions/checkout@v2
-      with:
-        fetch-depth: 0
-
-    - name: Derive appropriate SHAs for base and head for `nx affected` commands
-      uses: nrwl/nx-set-shas@v2
-
     - name: Use Node.js
-      uses: actions/setup-node@v2
+      uses: actions/setup-node@v3
       with:
         node-version: ${{ inputs.node-version }}
 
-    - uses: actions/cache@v2
+    - name: Get Yarn cache directory
+      id: yarn-cache-dir
+      shell: bash
+      run: echo "dir=$(yarn cache dir)" >> $GITHUB_OUTPUT
+
+    - name: Dependencies and Cypress cache
+      uses: actions/cache@v3
       id: yarn-cache
       with:
         path: |
-          node_modules
+          ${{ steps.yarn-cache-dir.outputs.dir }}
           ~/.cache/Cypress
-        key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+        key: ${{ runner.os }}-node-${{ inputs.node-version }}-yarn-${{ hashFiles('**/yarn.lock') }}
         restore-keys: |
-          ${{ runner.os }}-yarn-
+          ${{ runner.os }}-node-${{ inputs.node-version }}-
 
     - name: Install dependencies
-      if: steps.yarn-cache.outputs.cache-hit != 'true'
       shell: bash
```

Comment:
I don't get why this should be better. Now the installation & post-install hooks need to happen every time.

Reply:
One issue before was that the Node.js version wasn't taken into account. This causes problems because the `docs` project uses Node.js v16 while the rest use Node.js v14. Natively compiled modules are different between different operating systems and different versions of Node.js so they shouldn't be reused across operating systems or major Node.js versions.

In general, caching the compiled output in `node_modules` rather than the global package cache isn't recommended as some tools such as NGCC write to `node_modules`, sometimes in non-deterministic ways which can cause the fields to be corrupted, go stale, or out of sync.

There are other issues with using `node_modules` as some versions of some package managers delete files in `node_modules` as part of determining dependency graphs and linking dependencies.

Reproducibility and determinism should be preferred over speed if we have to choose between the two. Our docs CI builds have been flaky, likely because we have been caching the `node_modules` directory instead of the Yarn cache directory and sharing it across Node.js versions 14 and 16.

Ideally, we would just use the `cache: yarn` parameter of `actions/setup-node`, in fact I suggest we do that to decrease maintenance but we would still have to cache the Cypress cache directory manually by using `actions/cache`.